### PR TITLE
TINY-8747: Added file structure option to support legacy output

### PR DESCRIPTION
**Pull request by MitchCTiny**
Related Ticket: TINY-8747

Description of Changes:
* Added `structure` option with default value `flat` for all files output to same folder, or `legacy` for nested folder output

Pre-checks:
* [x] Changelog entry added
* [x] package.json version bumped (if first change of new major/minor)
* [x] Tests have been added (if applicable)

### CODE REVIEW

**Comment by lnewson** [comment]
@MitchCTiny it looks like you missed changing the default GitHub PR title here, so can you please correct that so it's something usable (especially since it'll be used in the final commit message).

File: CHANGELOG.md
Comment (lnewson) [inline CHANGELOG.md:11]: An `Added` should never be in a patch release as that goes against semver... it needs to be in either a minor or major release, so you'll need to bump the version to 0.3.0 or similar.

File: README.md
Comment (lnewson) [inline README.md:3]: h1 -> h3 headings are not semantically correct. It looks like these should be h2 (`##`) headings instead.

File: src/lib/api.ts
Comment (lnewson) [inline src/lib/api.ts:10]: This isn't the right place to put this, as it's not concerned with the builder/parsed data which is what the `Api` class entails (it's not named the best sorry). So instead we should be passing this through to the template renderer/exporter directly instead of this third-party approach.

File: src/lib/cli.ts
Comment (lnewson) [inline src/lib/cli.ts:15]: This doesn't make sense, as the rest of the existing templates do not output a `flat` structure... I don't think we can specify the default here as it will vary based on the template default.

File: src/lib/moxiedoc.ts
Comment (lnewson) [inline src/lib/moxiedoc.ts:100]: As noted previously, this shouldn't be done like this... we should be passing this to the `Exporter` below instead and then passing it to the template render functions.

**Review by lnewson — CHANGES_REQUESTED** [review]
Still need to review the antora template changes sorry, so just submitting what I have so far.

File: src/templates/antora/antora.converter.ts
Comment (lnewson) [inline src/templates/antora/antora.converter.ts:90]: This return type is incorrect... you're losing type information by doing this.
```suggestion
const getFilePath = (structure: string): (name: string) => string =>
```

File: src/templates/antora/antora.converter.ts
Comment (lnewson) [inline src/templates/antora/antora.converter.ts:85]: `name.split('.').slice(0, -1)` seems to be duplicated a couple times and looks to be related to the legacy logic... can we please pull that out to a separate function to make it clearer what it's doing?

File: src/templates/antora/antora.converter.ts
Comment (lnewson) [inline src/templates/antora/antora.converter.ts:93]: The `title` argument never seems to be used in any call. Is it necessary?

File: src/templates/antora/antora.converter.ts
Comment (lnewson) [inline src/templates/antora/antora.converter.ts:94]: This is variable mutation and it's not something we like to do at Tiny. So this should declare a new variable, though as noted in my previous comment `title` is never passed so we could just remove it instead so it becomes:
```ts
const title = name.split('.').slice(-1).join('');
```

File: src/templates/antora/template.ts
Comment (lnewson) [inline src/templates/antora/template.ts:44]: This needs the `description` and `keywords` added to it like it did in the old `tinymcenext` template: https://github.com/tinymce/moxiedoc/blob/39b04d910f4e85b9026406d26276286d9a2dd749/src/templates/tinymcenext/member.handlebars#L4-L5

File: src/templates/antora/template.ts
Comment (lnewson) [inline src/templates/antora/template.ts:303]: This doesn't look correct as the old system never generated a file like this. It should just be `index.adoc` in the correct directory instead.

File: src/templates/antora/template.ts
Comment (lnewson) [inline src/templates/antora/template.ts:179]: Why was this removed? The old version was more readable and it could have been updated to delegate to the `AntoraTemplate.getFilePath` Instead of inlining it like this.

File: src/templates/antora/antora.converter.ts
Comment (lnewson) [inline src/templates/antora/antora.converter.ts:371]: This logic probably should be pulled out into a utility module or similar if we now need it in multiple places.

File: src/templates/antora/antora.converter.ts
Comment (lnewson) [inline src/templates/antora/antora.converter.ts:79]: For consistency we maybe should rename this to include "flat" since that's the structure it generates.

File: src/templates/antora/template.ts
Comment (lnewson) [reply src/templates/antora/template.ts:44]: This probably should be using the `member.handlebars` template to generate the initial file like we do for other files.

File: src/templates/antora/template.ts
Comment (lnewson) [inline src/templates/antora/template.ts:51]: This here needs to generate more than just a xref... look at the old pages and you'll see it has a small little description as well: https://www.tiny.cloud/docs/api/tinymce.dom/

Comment (lnewson) [reply src/templates/antora/template.ts:51]: I believe https://github.com/tinymce/tinymce-docs/blob/main/_includes/index.html is how it used to generate the old "card" so that should give some indications on how to replicate that.

File: src/templates/antora/antora.converter.ts
Comment (MitchCTiny) [reply src/templates/antora/antora.converter.ts]: new function name `getNamespaceFromFullName`

File: src/templates/antora/template.ts
Comment (MitchCTiny) [reply src/templates/antora/template.ts]: Done

**Comment by MitchCTiny** [comment]
The docs branch [https://github.com/tinymce/tinymce-docs/tree/feature/5/TINY-8747] (feature/5/TINY-8747) has been updated for testing this PR. Just build moxiedoc on this branch, update the dist location in the `api-reference.sh` script, and run `yarn build && yarn serve` and go to localhost:4000 to see the changes to the API reference.

**Comment by ephox-mogran** [comment]
I don't know enough about moxiedoc (read: anything) to review it, sorry.

**Review by ephox-mogran — COMMENTED** [review]
Not familiar enough with it to review.

**Comment by lnewson** [comment]
> The docs branch `https://github.com/tinymce/tinymce-docs/tree/feature/5/TINY-8747` has been updated for testing this PR. Just build moxiedoc on this branch, update the dist location in the `api-reference.sh` script, and run `yarn build && yarn serve` and go to localhost:4000 to see the changes to the API reference.

@MitchCTiny this doesn't look updated to me? It's still using a version from way before you made the recent changes: https://github.com/tinymce/tinymce-docs/blob/feature/5/TINY-8747/package.json#L33

**Comment by MitchCTiny** [comment]
> > The docs branch `https://github.com/tinymce/tinymce-docs/tree/feature/5/TINY-8747` has been updated for testing this PR. Just build moxiedoc on this branch, update the dist location in the `api-reference.sh` script, and run `yarn build && yarn serve` and go to localhost:4000 to see the changes to the API reference.
> 
> @MitchCTiny this doesn't look updated to me? It's still using a version from way before you made the recent changes: https://github.com/tinymce/tinymce-docs/blob/feature/5/TINY-8747/package.json#L33

The `api-reference.sh` script uses a locally built version of the moxiedoc, I didn't think 0.3 was publishable until these changes were approved but if that's necessary it can be done.

**Comment by lnewson** [comment]
> > > The docs branch `https://github.com/tinymce/tinymce-docs/tree/feature/5/TINY-8747` has been updated for testing this PR. Just build moxiedoc on this branch, update the dist location in the `api-reference.sh` script, and run `yarn build && yarn serve` and go to localhost:4000 to see the changes to the API reference.
> > 
> > 
> > @MitchCTiny this doesn't look updated to me? It's still using a version from way before you made the recent changes: [https://github.com/tinymce/tinymce-docs/blob/feature/5/TINY-8747/package.json#L33](https://github.com/tinymce/tinymce-docs/blob/feature/5/TINY-8747/package.json?rgh-link-date=2022-08-23T00%3A20%3A27Z#L33)
> 
> The `api-reference.sh` script uses a locally built version of the moxiedoc, I didn't think 0.3 was publishable until these changes were approved but if that's necessary it can be done.

We should never ever be relying on a local build... if that's been changed then it needs to be corrected. Also remember that beehive-flow will publish prerelease versions for every CI build it does. That's what's currently being used and you can upgrade by doing something like `yarn upgrade "@tinymce/moxiedoc@feature-TINY-8747"`

**Comment by MitchCTiny** [comment]
> > The `api-reference.sh` script uses a locally built version of the moxiedoc, I didn't think 0.3 was publishable until these changes were approved but if that's necessary it can be done.
> 
> We should never ever be relying on a local build... if that's been changed then it needs to be corrected. Also remember that beehive-flow will publish prerelease versions for every CI build it does. That's what's currently being used and you can upgrade by doing something like `yarn upgrade "@tinymce/moxiedoc@feature-TINY-8747"`

Excellent, thank you. Updated the docs branch now.

File: README.md
Comment (lnewson) [inline README.md:23]: Is this still correct?

File: CHANGELOG.md
Comment (lnewson) [inline CHANGELOG.md:12]: Likewise should we be mentioning `default` here since it varies between templates? Also it doesn't look like `flat` is an option anymore?

File: src/templates/antora/template.ts
Comment (lnewson) [inline src/templates/antora/template.ts:192]:
```suggestion
  navPages.forEach(addPage);
```
This can be simplified/reduced to just this.

File: src/templates/antora/util.ts
Comment (lnewson) [inline src/templates/antora/util.ts:230]: This only works for legacy structures by the looks so maybe we should rename it as such? e.g `generateLegacyIndexPages`

File: src/templates/antora/util.ts
Comment (lnewson) [inline src/templates/antora/util.ts:76]: This path is incorrect and introduces a bug because of that... atm the JSON content is being transferred to Antora which we don't want. If you look at the old version this specific path should be hardcoded to `_data/api/json` instead of using the `BASE_PATH` since it's not Antora content.

- [ ] Fix bug with JSON output being placed in the wrong location

File: src/templates/antora/util.ts
Comment (lnewson) [inline src/templates/antora/util.ts:16]: Is this correct? It should only be able to output `json` and `adoc` to my knowledge.

**Review by lnewson — APPROVED** [review]
Good work on the cleanup since my last review @MitchCTiny, this looks much better 👍

**Comment by lnewson** [comment]
@MitchCTiny actually one question... the index pages that are generated for the legacy content. Are we sure that's correct? I thought the idea was to mirror what was there previously (i.e. have a 2 column table with each cell having the heading and description). The new version does **not** do that.

- [ ] Fix index pages so they behave the same as they did in 5.x

**Comment by lnewson** [comment]
To clarify the previous comment further, here's what your version looks like:
image
and here's what it's supposed to look like:
image

Now obviously we won't get the styling the same, but the structure should be the same where it looks like cards.

File: README.md
Comment (MitchCTiny) [reply README.md:23]: Nope, fixing to be either 'default' or 'legacy'

**Comment by lnewson** [comment]
As a tip for the card style layout, take a look at what we did in the TinyMCE 6 docs to replicate this. For example: https://github.com/tinymce/tinymce-docs/blob/staging/docs-6/modules/ROOT/partials/index-pages/opensource-plugins.adoc

File: src/templates/antora/util.ts
Comment (MitchCTiny) [reply src/templates/antora/util.ts:16]: Yep, reverting the change.

File: src/templates/antora/util.ts
Comment (MitchCTiny) [reply src/templates/antora/util.ts]: Done

File: src/templates/antora/template.ts
Comment (MitchCTiny) [reply src/templates/antora/template.ts]: Done

File: CHANGELOG.md
Comment (MitchCTiny) [reply CHANGELOG.md]: Updated

File: src/templates/antora/util.ts
Comment (MitchCTiny) [reply src/templates/antora/util.ts]: Fixed

**Comment by MitchCTiny** [comment]
> As a tip for the card style layout, take a look at what we did in the TinyMCE 6 docs to replicate this. For example: https://github.com/tinymce/tinymce-docs/blob/staging/docs-6/modules/ROOT/partials/index-pages/opensource-plugins.adoc

Ahh that's perfect, I wasn't sure how to go about that card visual style so I will implement that now.

File: CHANGELOG.md
Comment (lnewson) [inline CHANGELOG.md:9]:
```suggestion
## 0.3.0 - 2022-08-23
```
This date should also be updated given the difference (if it was only a couple days it'd be fine but over a month is too much).

File: CHANGELOG.md
Comment (lnewson) [inline CHANGELOG.md:12]:
```suggestion
- New `structure` option for specifying `legacy` nested structure for antora docs.
```
The last part here just didn't make sense. Feel free to come up with a different solution though.

Comment (MitchCTiny) [reply CHANGELOG.md:12]: Happy with yours

File: src/templates/antora/util.ts
Comment (lnewson) [inline src/templates/antora/util.ts:207]: This isn't the correct code style we use. It should either be on a single line or wrapped like this if needed:
```suggestion
const legacyIndexToAdoc = (
  namespace: NavFile,
  template: HandlebarsTemplateDelegate,
  descriptions: Record<string, string>,
  structure: ExportStructure
): string => {
```

Comment (MitchCTiny) [reply src/templates/antora/util.ts:207]: Ah I was wondering about that, I'll fix it up.

File: src/templates/antora/util.ts
Comment (lnewson) [inline src/templates/antora/util.ts:240]: Likewise here for the code style
```suggestion
const generateLegacyIndexPages = (
  indexPage: NavFile,
  sortedTypes: Type[],
  memberTemplate: HandlebarsTemplateDelegate,
  structure: ExportStructure
): PageOutput[] => {
```

**Pull request merged**
